### ISSUE TITLE
markup needed for display usefulness

### ISSUE THREAD

**Issue by funderburkjim** [sanskrit-lexicon/ApteES#9]
Yesterday I happened to use AE dictionary to find Sanskrit words for the English word 'last'.
I find the display to be ugly, awkward and hard to use.

Maybe some markup to ae.txt  will make AE more useful.  This issue to explore that idea.

Here's the display that I found objectionable:

<img width="306" alt="image" src="https://github.com/sanskrit-lexicon/ApteES/assets/6393033/35b9dde3-5d0b-4615-85be-eb6a7a7d8398">

**Comment by funderburkjim** [comment]
## cleanup begun.
It was soon clear that the markup of ae.txt needs much improvement before focusing on display improvements.  
[temp_ae_09.zip](https://github.com/sanskrit-lexicon/ApteES/files/15326823/temp_ae_09.zip) is the current state of the digitization.
I can see several additional things to do, such as suggested by the list of quotes [regex_quote_09.txt](https://github.com/sanskrit-lexicon/ApteES/blob/master/issues/issue9/1/regex_quote_09.txt).

Request @Andhrabharati to take a look at this 09 version, and provide feedback regarding things he notices that need further attention.

**Comment by Andhrabharati** [comment]
Yes, I could see more improvements/changes still; but I would first suggest looking for a mechanism to "eliminate" the blank lines within the body (i.e., between the L and LEND lines). I think these are retained (as of now) to keep the no. of lines the same as inthe original cdsl file.

**Comment by Andhrabharati** [comment]
I have opted to mark the blank lines within the body of the entries with Ⓑ; this is counted to 2012 instances at the end.

> 1. replace "\n\n<" with "\nⒷ\n<"	11417
> 2. replace "\n\n[" with "\nⒷ\n["	11
> 3. replace "\n\n(" with "\nⒷ\n("	42
> 4. replace "\n\n{" with "\nⒷ\n{"	511
> 5. replace "\n\n‘" with "\nⒷ\n‘"	90
> 6. replace "\n\n-" with "\nⒷ\n-"	7
> 7. replace "\n\n&" with "\nⒷ\n&"	6
> 8. replace "\n\n," with "\nⒷ\n,"	1
> 9. replace "\n\n([a-z])" with "\nⒷ\n\1"	14 [regex]
> 10. replace "\n.\n" with ".\nⒷ\n"	1211 + 35 + 2
> 11. replace "\n\nⒷ" with "\nⒷ\nⒷ"	47 + 20
> 12. replace "Ⓑ\n<L>" with "\n<L>"

Then looked for single blank between `<LEND>` line and the next `<L>` line--

> 1. Two entries L-4878 & L-5159 have two blank lines preceding the `<L>` line.
> 2. One entry L-5137 has no preceding blank line.
> 3. At 6  places the [Pagexxx] (194, 242, 249, 375, 484, 501) is relocated to the prev. line of preceding `<H>` line.

**Comment by Andhrabharati** [comment]
I've come up with a simple (and great!) idea to correct the AE text, by putting it into a "structured form".
[Have used the 🞄 (U+1F784) as the marker to join the lines of the entries into a single line first; thereafter had split the same as per the envisaged structure.]

Few sample entries--
![image](https://github.com/sanskrit-lexicon/ApteES/assets/75209130/d70c9b9d-bc20-4668-b8ac-4bb03ddf289d)

![image](https://github.com/sanskrit-lexicon/ApteES/assets/75209130/693d9316-65a5-47cd-804f-34161c1c0e1b)

Got quite many corrections in the process [**2336 inline differences in 1972 changed lines**], and here is a new entry--
lines 88502-3

"`{@-ed, -y.@} {%a.%} {#reKAMkita, BaMgavat#}.\n`
`{@rist@}, {%s.%} {#maRibaMDaH-DanaM, prakozWaH#}.`"

to change as

"`{@-ed, -y@}, {%a.%} {#reKAMkita, BaMgavat#}.\n`
`<LEND>\n`
`\n`
`<L>11310.1<pc>499<k1>wrist<k2>wrist\n`
`{@Wrist@}, {%s.%} {#maRibaMDaH-DanaM, prakozWaH#}.`"

and here is a missed line--
line 12576:

"`{#tulya-samAna#} -{%a.%}  {#BU#}. {@-ence@}, {%s.%} {#saMnipAtaH#},`"

to change as

"`{#tulya-samAna-#} -{%a.%} {#BU, yuj#} pass. {#anuvft#} 1 A {#saMvad#} 1 P,\n`
`{#tulya-samAna#} {@-ence@}, {%s.%} {#saMnipAtaH#},`"

**Comment by funderburkjim** [comment]
We are thinking in a similar way!  A structured form is the target, as your 'last' display form nicely exemplifies.  I had noticed the '@' elements as structure divisions.  But think the POS elements you included look useful also.  I notice you added `{@1@}` under last,  which for some reason Apte always omits  (but there are 5 erroneous `{@1@} A.` in cdsl).

 >> Got quite many corrections ...

Would you share these corrections? (or maybe share your entire version?)

------------------------------------------

One other aspect we should aim for is the expansion of abbreviations, at these two types 
* `‘l. ing for many days’` -> `‘lasting for many days’`  under hw 'last'
* `{@-ence@}`  ->  `{@coincidence@}   (under hw 'coincide').

**Comment by Andhrabharati** [comment]
How should the 'filling' of HWs and the expansion of local abbr.s be done?

Straight replacement with the full form or with some 'notation'?

**Comment by Andhrabharati** [comment]
> but there are 5 erroneous `{@1@} A.` in cdsl

There are 35 instances of `{@[0-9]@} [APU]` in the cdsl text that indicate the conjugation terms, and the bold marking `{@x@}` is to be removed from these.

**Comment by funderburkjim** [comment]
I'm not sure what final notation should be for the expansions of 
* {@-X@} and 
* ‘... x. ...’

What about first making a separate table with 3 columns, such as this for 'abandon' ?
```
<L>2{@Abandon@} : {@-ed.@} : {@Abandoned@}
<L>2{@Abandon@} : ‘he was a. to his fate’ : ‘he was abandoned to his fate’
<L>2{@Abandon@} : {@-ment@} : {@Abandonment@}
<L>2{@Abandon@} : ‘a. of worldly connections’ : ‘abandonment of worldly connections’
```
A program could then apply this table to ae.txt.  At this point of application, we could either use the column 3 replacement as shown, or we could introduce some notation if that seems better.

-------------------------------------

> There are 35 instances of {@[0-9]@} [APU] ...

Agree.

**Comment by Andhrabharati** [comment]
Started with tagging the abbr. lex. & ls entities and finished.

Marking HW fillings with [xx] notation and reached about 3/4th work.

Marking the local contractions with (xxx) notation and did about 1/3rd work. Also found many strings not in quotes, and added the quote marks to those.

**Comment by funderburkjim** [comment]
I've also made numerous corrections and changes.  
My latest version is [temp_ae_12.zip](https://github.com/sanskrit-lexicon/ApteES/files/15370773/temp_ae_12.zip)

I'm not against removing blank lines and intra-entry lines, but have not done so yet.

Not sure how we can coordinate our efforts.

**Comment by gasyoun** [comment]
@funderburkjim some notation if that seems better -- indeed.

**Comment by funderburkjim** [comment]
A wrong headword:
`<L>7562<pc>328<k1>pained<k2>pained`  this headword removed; it is part of 
`<L>7561<pc>327<k1>pain<k2>pain`.

**Comment by funderburkjim** [comment]
@Andhrabharati  please check your change mentioned above under 'coincide'.  Here is my current:
```
<L>1915<pc>063<k1>coincide<k2>coincide
{@Coincide@}¦, {%v. i.%} {#ekIBU#} 1 P, {#saMgam#} 1 A, {#saM-sannipat#}
1 P; {#saMDA-saMyuj#} -pass. {@2@} {#saMman#}
4 A, {#ekacittIBU#}; {%See%} {@Agree@} {@3@} {#anurUpa#}
{%a.%} {#BU, yuj#} {%pass.%} {#anuvft#} 1 A {#saMvad#} 1 P,
{#tulya-samAna#} -{%a.%}  {#BU#}. {@-ence@}, {%s.%} {#saMnipAtaH#},
{#saMgamaH; saMyogaH#}. {@2@} {#saMmati#} {%f.%}, {#EkamatyaM, saMDiH#},
{#ekacittatA#} {@3@} {#saMvAdaH, anurUpatA, AnuguRyaM#}.
{@4@} {#dEvayogaH-GawanaM#}; ‘be an unlucky c.’
{#durdEvAt#} {@-ent@}, {%a.%} {#saMpAptin#}. {@2@} {#samakAlIna#}.
{@3@} {#avirudDa, anurUpa, saMvAdin#}.
<LEND>
```

**Comment by Andhrabharati** [comment]
Yes, I did something wrong here-- while checking and reporting!

My files [**present** and **older**] also have the same content.

**Comment by Andhrabharati** [comment]
Add a new entry--

```
<L>7636.1<pc>330<k1>Parallelopiped<k2>Parallelopiped
{@Parallelopiped@}¦, ✦{%s.%}	⧫{#samAMtaraKAtaH#}.
<LEND>
```

**Comment by Andhrabharati** [comment]
Here is my intermittent file for your study/comparison, @funderburkjim --
[temp_ae_09D.zip](https://github.com/sanskrit-lexicon/ApteES/files/15392949/temp_ae_09D.zip)

This differs by 6 lines with your ver. 12 above (at my entries 7636.1 & 11310.1).

And these are the replacements to get the file into the "structure" I had mentioned above--
" 🞄\n"	->	" 🞄"
"🠚⧫"	->	"\t⧫"
"🠚✦"	->	"\t✦"	;; for entries or `<div/>`s with body portion
"⇨∆"	->	"\n\t∆"	;; for entries or `<div/>`s without body portion
"Ⓝ"	->	"\nⓃ"
"Ⓝ"	->	"`<div/>`"

I've seen that you had erred in hyphenation at many places, both in English words as well as in Skt. words.
My next task would to be correct those, before proceeding with expansion of the rest of the local contractions (mostly within quotes).

**Comment by funderburkjim** [comment]
Found one more like 'Bacchanal', etc.
L=8669 Reception to be merged into 8668 Receptacle.

**Comment by funderburkjim** [comment]
`<L>7636.1<pc>330<k1>Parallelopiped<k2>Parallelopiped`

k1,k2 should be lower-case in metaline (per cdsl convention in ae.txt)

`<L>7636.1<pc>330<k1>parallelopiped<k2>parallelopiped`

**Comment by funderburkjim** [comment]
## change of extended ascii temporary markup
Emacs does not display 3 of the characters.  Thus I am changing to other characters of similar appearance which are shown in Emacs.  I am using Consolas font in Emacs (one of the mono-spaced fonts).


```
AB:
⧫  (\u29eb) 32622 := BLACK LOZENGE
🞄  (\u1f784) 43062 := BLACK SLIGHTLY SMALL CIRCLE
🠚  (\u1f81a) 53047 := HEAVY RIGHTWARDS ARROW WITH EQUILATERAL ARROWHEAD
CDSL:
♦  (\u2666) 32622 := BLACK DIAMOND SUIT
•  (\u2022) 43062 := BULLET
➔  (\u2794) 53047 := HEAVY WIDE-HEADED RIGHTWARDS ARROW

```

Here is the first line, in both forms.
```
AB:
@A@}, {@An@}¦,🠚✦🠚⧫(As an article) not <ab>ex.</ab>; ‘a man’ 🞄
CDSL:
{@A@}, {@An@}¦,➔✦➔♦(As an article) not <ab>ex.</ab>; ‘a man’ •
```

[temp_ae_09D_ea.txt](https://github.com/sanskrit-lexicon/ApteES/files/15395717/temp_ae_09D_ea.txt)  extended ascii in temp_ae_09D/


[temp_ae_09D_cdsl_ea.txt](https://github.com/sanskrit-lexicon/ApteES/files/15395723/temp_ae_09D_cdsl_ea.txt) in cdsl .

**Comment by funderburkjim** [comment]
## Version 15:  'sub-entries'
* [temp_ae_15.zip](https://github.com/sanskrit-lexicon/ApteES/files/15462686/temp_ae_15.zip) contains a new version.
* [temp_ae_ab2.zip](https://github.com/sanskrit-lexicon/ApteES/files/15462689/temp_ae_ab2.zip) contains similar format

Both of these are powered by [temp_ae_09D.zip](https://github.com/sanskrit-lexicon/ApteES/files/15392949/temp_ae_09D.zip)  from above.

Essentially, `Ⓝ{@[Abandon]ed@},➔` under Abandon is turned into a new 'sub-entry'.
The main entry Abandon is noted as `<e>1` on the metaline, and the sub-entries are noted by `<e>2`; this is similar to MW coding at cdsl.

What do you think of this approach, @Andhrabharati ?

------------------------------------------------------
The GRA/BOP form for 'k2' is adopted in the metaline.  'Alternate' headwords will make use of this.

It is an open question what 'k1' should be for 'phrases'. version 15 has k1='set_about' for the 'set about' sub-entry based on 'set'.

------------------------------------------------------
The next step will be to add the POS (part of speech) markup to version 15. This is already present in the form `✦{%X.%}➔`  in version ab2.

**Comment by Andhrabharati** [comment]
> Essentially, `Ⓝ{@[Abandon]ed@},➔` under Abandon is turned into a new 'sub-entry'. The main entry Abandon is noted as `<e>1` on the metaline, and the sub-entries are noted by `<e>2`; this is similar to MW coding at cdsl.
> 
> What do you think of this approach, @Andhrabharati ?

Yes, it is a proper way to go; but I feel using `<e>3` would've been appropriate.
[MW has the composite & derived words in contracted form, that we denote as `<e>3`, whereas the `<e>2` notation is used for full HWs (but in IAST spelling).]

> It is an open question what 'k1' should be for 'phrases'. version 15 has k1='set_about' for the 'set about' sub-entry based on 'set'.

I wonder why this ('_' for phrases) has been **discontinued in MW now**; it used to be there when I started interacting with CDSL (few years back).

> The next step will be to add the POS (part of speech) markup to version 15. This is already present in the form `✦{%X.%}➔` in version ab2.

I have finished expanding the contracted forms [in (xxx) notation, as mentioned above]; probably Jim should start afresh with this revised file, which also contains numerous other corrections-- including punctuation, HWs (basic as well as derived Ⓝ), POS, ab & ls markings, ... ...

**If Jim agrees for this, I shall upload my latest file.**

- However, I did not put my mind in "resolving" these expansions as (a) full words, (b) hyphenated words or (c) phrases [this resolution was done "properly" in case of the derived HWs in the prev. 09D file that I had uploaded]. The main reason being that the ae.txt and print (1920) contain punctuation errors all over.

**This completes the markup phase from my side.**

The remaining work is mostly correcting [textual errors](https://github.com/sanskrit-lexicon/ApteES/issues/9#issuecomment-2124985800), incl. hyphenation errors in Jim's initial work.
[I am just wondering if this should be taken up!!]

`------------------------------------------`
BTW, I had seen that to appropriately conclude on the "text" of AE, the earlier ed. of 1893 helps a lot [the 3rd ed. of 1920 having introduced quite many errors in the print.]

**Comment by Andhrabharati** [comment]
> The GRA/BOP form for 'k2' is adopted in the metaline. 'Alternate' headwords will make use of this.

It is seen that 326 "grouped" entries (out of 832) have only a single element in k2-field [**in the temp_ae_15.txt**]!!

And as done in GRA, I'd suggest that the splitting of grouped entries into individual k2-entries be done in the .xml file, leaving the .txt file in a bare-form close to the print.

**Comment by funderburkjim** [comment]
> If Jim agrees for this, I shall upload my latest file.

Yes, agree. Please do upload.  Also mention any formal differences from 09D version.  
You might incorporate the few [changes](https://github.com/sanskrit-lexicon/ApteES/blob/master/issues/issue9/1/ABmark/change_ab_1_1a.txt) I made to your file.

> 326 "grouped" entries (out of 832) have only a single element in k2

Perhaps there is some form of indicating multiple headwords that I missed. If it is easy, upload that list of 326.

> taking up textual errors

Certainly needs to be done sometime.  Up to you as to when the time is best for this.

**Comment by Andhrabharati** [comment]
> > 326 "grouped" entries (out of 832) have only a single element in k2
> 
> Perhaps there is some form of indicating multiple headwords that I missed. If it is easy, upload that list of 326.

You can get these by the regex `<k2>[^,]*$\n(.*),(.*)¦`

**Comment by Andhrabharati** [comment]
> Yes, agree. Please do upload. Also mention any formal differences from 09D version.
> You might incorporate the few [changes](https://github.com/sanskrit-lexicon/ApteES/blob/master/issues/issue9/1/ABmark/change_ab_1_1a.txt) I made to your file.

I shall go through your changes in the above file, and upload my revised file before I retire for the day.

**Comment by Andhrabharati** [comment]
Here is my latest file, @funderburkjim --
[temp_ae_09E.zip](https://github.com/sanskrit-lexicon/ApteES/files/15474518/temp_ae_09E.zip)

Changes wrt 09D:

**A. tag-replacements** (to make these less obtrusive in free reading)
`<ab>`	->	⁅
`</ab>`	->	⁆
`<ls>`	->	〔
`</ls>`	->	〕

**B. Notes:**
1. I have opted to retain the ligatures  æ and œ.
2. k2-field is corrected/populated as per the header portion, but k1-field is mostly left untouched.
3. {%xxx.%}  is changed as `<lex>...</lex>` and {%yyy%} as `<i>...</i>`
4. `<lex>...</lex>` is rendered in italics, whereas the `<ab>...</ab>` is in normal-face, in the print.
5. %...% is the "padded" text as per the context, which is not in the print. [at 5 places]
6. Added the homonym numbers ⒈ & ⒉ wherever deemed necessary.

**Hope you would face no issues in using my revised file.** You might need to do the [extended ascii changes](https://github.com/sanskrit-lexicon/ApteES/issues/9#issuecomment-2123499777) in this 09E file.

**Comment by funderburkjim** [comment]
Thank you. I'll begin working with your file on the next phase, probably in a couple of days.

**Comment by Andhrabharati** [comment]
Pl. change `???` at `<L>3503` as 94.

**Comment by Andhrabharati** [comment]
> Thank you. I'll begin working with your file on the next phase, probably in a couple of days.

You may pl. ping me once before you start this, @funderburkjim !

I should be able to give you my next phase of work, with good many corrections (recognized by-way-of "reformatting & re-grouping" of the text strings).

I've also started correcting the "residual" hyphenations carried-over from your "base" work posted at the top of this issue.

**Comment by funderburkjim** [comment]
I'm sidetracked by the mw backlog work.  Hope to return to this soon.

**Comment by Andhrabharati** [comment]
> Noticed **good many number of corrections** in the Devanagari text!!
> 
> It is a better idea to somehow compare with the data in use at our own site (andhrabharati.com), than to do a fresh proofing of cdsl text. [We have done the full proofing those initial days itself.]
> 
> But it needs some "introspection" as to how to compare the text with almost **poles-apart** structures; so keeping it on hold till the end of the current exercise of "marking" the cdsl text.

... ... ... ... ... ...

> The remaining work is mostly correcting [textual errors](https://github.com/sanskrit-lexicon/ApteES/issues/9#issuecomment-2124985800), incl. hyphenation errors in Jim's initial work.
> [I am just wondering if this should be taken up!!]

I just envisaged a way to correlate the Devanagari strings of the CDSL text (duly tagged now in multiple types), and the proofed text of andhrabharati.com; and will be taking up the task in next few days.
[Estimating about 4-5 days time to finish the work.]

**Comment by Andhrabharati** [comment]
The outcome of correlating with earlier AB file (of 2014) data--

- Over 1200 corrections made, ignoring the spaces and hyphens.
- The print having dual style (i.e. with & with-out) in hyphenating the (upasarga-)prefixed roots, a uniform way of "removing" the hyphens at the line-break has been followed in the text.
- Re-positioned a huge no. of (over 500 or so) example phrases (that are given at the "end of the entry", just before the start of the composite/derived HWs) to the respective places inside the body.

As the print (not scan!) quality [**I recall having mentioned sometime back, of having a good quality print & scan copies with me**] is quite bad (wrt punctuations etc.), another round of reading with the earlier (2nd) ed. of 1894 might weed out more errors in the text file.

@funderburkjim 
Are you ready to resume work on this, now?

And, would you like to have the file with CDSL style (with line-breaks in-tact), or in the proposed style (split as per the entry words)?

As the data is kind of mostly proofed and restructured now, retaining the line-breaks of the original digitisation is not so important now (just as in the MW text).

**Comment by funderburkjim** [comment]
>  Are you ready to resume?

@Andhrabharati   I think I should do the {{Lbody}} conversion for the 11 dictionaries ([Refer](https://github.com/sanskrit-lexicon/GRA/issues/34)) while the details  are at front of mind. Then, I hope to return to this issue.  Estimate a week.

**Comment by Andhrabharati** [comment]
Yes, doing a particular task across all the works, is what I also do many a times.

But, looking at the list of works coming under this, I should be "provided" a version in my "comma-separated k2" method, to continue (from the major works that I had done in PWG/pwk, VCP, SKD) which are yet to be shared with CDSL.

[I have done sizeable work in these also, as in MW.]

**Comment by Andhrabharati** [comment]
Even GRA, BEN & BUR are needed in the erstewhile "comma-separated k2" versions, to correlate my working in them.
[I did not touch ACC, CAE and SHS yet!!]

**Comment by funderburkjim** [comment]
@Andhrabharati  I plan to do the cdsl conversions from hwextra to Lbody for all the dictionaries.

Then, I'll aim to revise the Lbody-k2 conversion (as developed for MW [here](https://github.com/sanskrit-lexicon/MWS/issues/176)) so it can 
apply to any of the dictionaries.

**Comment by Andhrabharati** [comment]
Glad to know this; I am much obliged.

This puts me in a "comfortable mode" to work on the CDSL dictionaries.

**Comment by Andhrabharati** [comment]
@funderburkjim 

This is my final reminder on the topic!
[You had asked me to do some [expansions](https://github.com/sanskrit-lexicon/ApteES/issues/9#issuecomment-2118028795) that has resulted into a major revamp of the whole text.]

Are you still in a position to take a look at [**and of course, to work with**] my (latest) version [**not yet posted**, awaiting your response to "resume" the work] of this AE?

**You may browse through the above posts in May-Aug. 2024, just to know what all I have done.**

**Comment by funderburkjim** [comment]
@Andhrabharati   I definitely would like to work with your latest version of AE in order to make a better cdsl version.   My intention is to take up work with AE when my work with AP(57) is at a good stopping point.
I hope your AE work will be available at that time.

**Comment by Andhrabharati** [comment]
That's good, noting that that one of my works won't "go waste"!

Let's wait then, for a reasonable closure time of AP57.

**Comment by funderburkjim** [comment]
@Andhrabharati   Ready (finally!) to begin work on AE.
Anticipate leaning heavily on your work.  Request you provide
your version that I should start with.

**Comment by Andhrabharati** [comment]
Great!

As you haven't responded about the version that I [asked](https://github.com/sanskrit-lexicon/ApteES/issues/9#issuecomment-2305025221), I thought I should give both the versions together and leave the choice to you--

[ae_10A_deva.txt](https://github.com/user-attachments/files/27414745/ae_10A_deva.txt)

[AE (New format).txt](https://github.com/user-attachments/files/27414747/AE.New.format.txt)

And, **you need to go through my above posts**, to know about the details of my file structuring.

**Comment by Andhrabharati** [comment]
BTW, I also urge you to take up another major task wrt my earlier works, namely marking and expanding the abbr.s in PWG, of whose the comparison is given below--

<img width="511" height="130" alt="Image" src="https://github.com/user-attachments/assets/1489b631-3006-4d5f-8289-b45172a120bd" />

–––––––––––
And, it may be recalled that my pwk version has been adopted into CDSL "as is" earlier, whose abbr. counts stand as below--

<img width="234" height="72" alt="Image" src="https://github.com/user-attachments/assets/1f7de1cf-d8ee-4826-8060-dbf60b87c2aa" />

**Comment by funderburkjim** [comment]
@Andhrabharati   Do your two files include changes to cdsl ae made since 2024 ?  

[readme_ae_history.txt](https://github.com/sanskrit-lexicon/ApteES/blob/master/issues/issue9/readme_ae_history.txt) might be relevant.

**Comment by Andhrabharati** [comment]
yes.

**Comment by Andhrabharati** [comment]
But you may crosscheck once randomly, if not fully.
